crates/storage/provider/src/providers/mod.rs

```diff
@@ -231,6 +240,60 @@ impl<DB: Database> WithdrawalsProvider for ShareableDatabase<DB> {
     }
 }
 
+impl<DB: Database> EvmEnvProvider for ShareableDatabase<DB> {
+    fn fill_env_at(&self, env: &mut Env, at: BlockId) -> Result<()> {
+        let hash = self.block_hash_for_id(at)?.ok_or(ProviderError::HeaderNotFound)?;
+        let header = self.header(&hash)?.ok_or(ProviderError::HeaderNotFound)?;
```

Comment:
unrelated to this PR: we should prob accept BlockId in `HeaderProvider::header` and `HeaderProvider::header_td`